### ISSUE TITLE
expand description field to allow json instead of just quoted-string and token

### ISSUE THREAD

**Issue by colinbendell** [w3c/server-timing#45]
A common use for the description field is to add additional details that can be used by other visualization or systems. However this is limiting because it does not allow the use of json values nor other name/value pair notations. I propose that instead of using RFC7230's quoted-string value we expect the value to be a json value.

For example, if the server wanted to include the attributes `request-id`, `o-bytes` and `product-id` in the desc field, in the following json:
```json
{
  "request-id": "8861050c",
  "o-bytes": 580970,
  "product-id": "e39141b",
  "cache": "hit"
}
```

To include this in the `desc` entry all the quotes must be escaped:
```
Server-timing: origin; dur=50; desc="{\"request-id\":\"8861050c\", \"o-bytes\": 580970, \"product-id\": \"e39141b\",\"cache\": \"hit\"}";
```

However, even this is still invalid because of the use of `{`, `}`, and `:`. Even characters such as `=`, ',' and ';'. This means that even if you wanted to create a pseudo name-value pair set you have to use a limited collection. 

A viable solution is:
```
Server-timing: origin; dur=50; desc="requestid_8861050c+obytes_580970+productid_e39141b+cache_hit";
```

This is now problematic because I have invented my own token syntax that is non standard and requires any consumer to be aware of my use of `_` and `+`.

To address these issues, I would propose that the `desc=` field should use json as a valid value. I would further propose that the spec require `desc=` to occur only once on a given `Server-Timing` entry, and be required to be the last value to as not to confuse the parser of with other tokens.

(Alternatively, the `desc=` could be expanded to allow for token, quoted-value or json but I don't know what the implications are for the parser implementation)

**Comment by cvazac** [comment]
`{` (x7b), `}` (x7d), and `:` (x3a) are all legal in quoted-strings, see [definition](https://tools.ietf.org/html/rfc7230#section-3.2.6) of `qdtext`:
```
     qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
```

I will add a test cases for this everywhere I can.

**Comment by colinbendell** [comment]
you are right, within dtext these are valid characters. I've updated the issue.

However, the desc requires all these double quotes which makes the value illegible outside of the context and requires post processing because of the extra superfluous characters. If we are concerned about the 5 characters of `duration` => `dur` then we should be likewise concerned about all the extra `\` characters

**Comment by colinbendell** [comment]
sure. but now it's not valid json (per RFC7159 is double quote `"`") and again consumers will not only have to know to `s/'/"/g` but now also `s/\\'/'/g`. the intention here is that the consumer (human or machine) should not have to know any prior parsing contract and the contents should be natural to digest.

**Comment by igrigorik** [comment]
Paging @mnot for guidance. Mark, what's the status of [structured headers](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html) proposal, and what's the best way to tackle what we have here at this point?

**Comment by mnot** [comment]
It isn't currently a goal for structured data to carry arbitrarily nested structured data, nor to carry JSON.

However, you could carry JSON inside a SH string, and it would be correctly escaped and un-escaped by SH tooling on each end. Yes, you'd have a bit of overhead with the escapes.

Why can't those attributes be put in other params? It doesn't seem like great practice to overload `desc` like this -- my understanding was that it was intended for presentation to humans (e.g., in devtools).
